Cgen always uses `thread_arg_`

### Describe the bug

<title>

### Reproduction Steps

```v
import time

pub type EventListener[T] = fn (T) !

@[heap]
struct Chan[T] {
	c chan T
}

pub type Check[T] = fn (T) bool

struct EventWaiter[T] {
	check ?Check[T]
	c     &Chan[T]
}

pub struct EventController[T] {
mut:
	id        int
	wait_fors map[int]EventWaiter[T]
	listeners map[int]EventListener[T]
}

fn (mut ec EventController[T]) generate_id() int {
	return ec.id++
}

@[params]
pub struct EmitOptions {
pub:
	error_handler ?fn (int, IError)
}

pub fn (mut ec EventController[T]) emit(e T, options EmitOptions) {
	for i, w in ec.wait_fors {
		mut b := false
		if w.check != none {
			b = (w.check or { panic(err) })(e)
		} else {
			b = true
		}
		if b {
			w.c.c <- e
			ec.wait_fors.delete(i)
			return
		}
	}
	mut ts := []thread{}
	for i, l in ec.listeners {
		ts << spawn fn [options] [T](f EventListener[T], j int, e T) {
			f(e) or {
				if g := options.error_handler {
					g(j, err)
				}
			}
		}(l, i, e)
	}
	ts.wait()
}

@[params]
pub struct EventWaitParams[T] {
pub:
	check   ?Check[T]
	timeout ?time.Duration
}

pub struct Awaitable[T] {
	id int
	timeout ?time.Duration
	c Chan[T]
mut:
	controller &EventController[T]
}

struct Zero[T] {
	e T
}

pub fn (mut a Awaitable[T]) do() ?T {
	defer {
		a.controller.wait_fors.delete(a.id)
	}
	if timeout := a.timeout {
		mut e := Zero[T]{}.e
		select {
			e = <-a.c.c {
				r := e
				return r
			}
			timeout.nanoseconds() {
				return none
			}
		}
	}
	return <-a.c.c

}


pub fn (mut ec EventController[T]) wait(params EventWaitParams[T]) Awaitable[T] {
	mut c := Chan[T]{}
	id := ec.generate_id()
	ec.wait_fors[id] = EventWaiter[T]{
		check: params.check
		c: &mut c
	}
	return Awaitable[T]{
		id: id
		timeout: params.timeout
		controller: unsafe { &mut ec }
		c: c
	}
}

struct Foo {}
struct Bar {}

fn main() {
	x := EventController[Foo]{}
	y := EventController[Bar]{}
}
```

### Expected Behavior

Successful compile and no errors

### Current Behavior

```v
PS D:\Games\Proekti\V\interactions> v run repro.v
repro.v:120:2: warning: unused variable: `x`
  118 |
  119 | fn main() {
  120 |     x := EventController[Foo]{}
      |     ^
  121 |     y := EventController[Bar]{}
  122 | }
repro.v:121:2: warning: unused variable: `y`
  119 | fn main() {
  120 |     x := EventController[Foo]{}
  121 |     y := EventController[Bar]{}
      |     ^
  122 | }
==================
C:/Users/mclr/AppData/Local/Temp/v_0/repro.10062483905514704926.tmp.c:18950: warning: assignment from incompatible pointer type
C:/Users/mclr/AppData/Local/Temp/v_0/repro.10062483905514704926.tmp.c:18952: error: cannot convert 'struct main__Bar' to 'struct main__Foo'
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.3 964798d

### Environment details (OS name and version, etc.)

```v
V full version: V 0.4.3 140c838.964798d
OS: windows, Microsoft Windows 11 Pro v22000 64-bit
Processor: 12 cpus, 64bit, little endian, 

getwd: D:\Games\Proekti\V\interactions
vexe: D:\Games\Proekti\V\v\v.exe
vexe mtime: 2023-12-10 23:17:27

vroot: OK, value: D:\Games\Proekti\V\v
VMODULES: OK, value: C:\Users\mclr\.vmodules
VTMP: OK, value: C:\Users\mclr\AppData\Local\Temp\v_0

Git version: git version 2.37.0.windows.1
Git vroot status: weekly.2023.48-85-g964798d5
.git/config present: true

CC version:
thirdparty/tcc status: thirdparty-windows-amd64 e90c2620
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -4236,11 +4236,12 @@ fn (mut c Checker) prefix_expr(mut node ast.PrefixExpr) ast.Type {
 		c.type_error_for_operator('-', 'numeric', right_sym.name, node.pos)
 	}
 	if node.op == .arrow {
-		if right_sym.kind == .chan {
+		raw_right_sym := c.table.final_sym(right_type)
+		if raw_right_sym.kind == .chan {
 			c.stmts_ending_with_expression(mut node.or_block.stmts)
-			return right_sym.chan_info().elem_type
+			return raw_right_sym.chan_info().elem_type
 		}
-		c.type_error_for_operator('<-', '`chan`', right_sym.name, node.pos)
+		c.type_error_for_operator('<-', '`chan`', raw_right_sym.name, node.pos)
 	}
 	return right_type
 }

--- a/vlib/v/gen/c/spawn_and_go.v
+++ b/vlib/v/gen/c/spawn_and_go.v
@@ -45,6 +45,7 @@ fn (mut g Gen) spawn_and_go_expr(node ast.SpawnExpr, mode SpawnGoMode) {
 			g.gen_anon_fn(mut expr.left)
 			g.writeln(';')
 			use_tmp_fn_var = true
+			name = g.anon_fn_cname(expr.left.decl.return_type, expr.left.decl.params.map(it.typ))
 		} else {
 			g.gen_anon_fn_decl(mut expr.left)
 			name = expr.left.decl.name

--- a/vlib/v/gen/c/utils.v
+++ b/vlib/v/gen/c/utils.v
@@ -113,6 +113,27 @@ fn (mut g Gen) fn_var_signature(return_type ast.Type, arg_types []ast.Type, var_
 	return sig
 }
 
+// generate anon fn cname, e.g. `anon_fn_void_int_string`
+fn (mut g Gen) anon_fn_cname(return_type ast.Type, arg_types []ast.Type) string {
+	ret_styp := c_name(g.typ(return_type))
+	mut sig := '${ret_styp}_'
+	for j, arg_typ in arg_types {
+		arg_sym := g.table.sym(arg_typ)
+		if arg_sym.info is ast.FnType {
+			func := arg_sym.info.func
+			arg_sig := g.anon_fn_cname(func.return_type, func.params.map(it.typ))
+			sig += 'fnon_fn_' + arg_sig
+		} else {
+			arg_styp := c_name(g.typ(arg_typ))
+			sig += arg_styp
+		}
+		if j < arg_types.len - 1 {
+			sig += '_'
+		}
+	}
+	return sig
+}
+
 // escape quotes for string
 fn escape_quotes(val string) string {
 	bs := '\\'

--- a/vlib/v/tests/generics_chans_select_test.v
+++ b/vlib/v/tests/generics_chans_select_test.v
@@ -1,0 +1,125 @@
+import time
+
+pub type EventListener[T] = fn (T) !
+
+@[heap]
+struct Chan[T] {
+	c chan T
+}
+
+pub type Check[T] = fn (T) bool
+
+struct EventWaiter[T] {
+	check ?Check[T]
+	c     &Chan[T]
+}
+
+pub struct EventController[T] {
+mut:
+	id        int
+	wait_fors map[int]EventWaiter[T]
+	listeners map[int]EventListener[T]
+}
+
+fn (mut ec EventController[T]) generate_id() int {
+	return ec.id++
+}
+
+@[params]
+pub struct EmitOptions {
+pub:
+	error_handler ?fn (int, IError)
+}
+
+pub fn (mut ec EventController[T]) emit(e T, options EmitOptions) {
+	for i, w in ec.wait_fors {
+		mut b := false
+		if w.check != none {
+			b = (w.check or { panic(err) })
+			e
+		} else {
+			b = true
+		}
+		if b {
+			w.c.c <- e
+			ec.wait_fors.delete(i)
+			return
+		}
+	}
+	mut ts := []thread{}
+	for i, l in ec.listeners {
+		ts << spawn fn [options] [T](f EventListener[T], j int, e T) {
+			f(e) or {
+				if g := options.error_handler {
+					g(j, err)
+				}
+			}
+		}(l, i, e)
+	}
+	ts.wait()
+}
+
+@[params]
+pub struct EventWaitParams[T] {
+pub:
+	check   ?Check[T]
+	timeout ?time.Duration
+}
+
+pub struct Awaitable[T] {
+	id      int
+	timeout ?time.Duration
+	c       Chan[T]
+mut:
+	controller &EventController[T]
+}
+
+struct Zero[T] {
+	e T
+}
+
+pub fn (mut a Awaitable[T]) do() ?T {
+	defer {
+		a.controller.wait_fors.delete(a.id)
+	}
+	if timeout := a.timeout {
+		mut e := Zero[T]{}.e
+		select {
+			e = <-a.c.c {
+				r := e
+				return r
+			}
+			timeout.nanoseconds() {
+				return none
+			}
+		}
+	}
+	return <-a.c.c
+}
+
+pub fn (mut ec EventController[T]) wait(params EventWaitParams[T]) Awaitable[T] {
+	mut c := Chan[T]{}
+	id := ec.generate_id()
+	ec.wait_fors[id] = EventWaiter[T]{
+		check: params.check
+		c: &mut c
+	}
+	return Awaitable[T]{
+		id: id
+		timeout: params.timeout
+		controller: unsafe { &mut ec }
+		c: c
+	}
+}
+
+struct Foo {}
+
+struct Bar {}
+
+fn test_generics_chans_select() {
+	x := EventController[Foo]{}
+	println(x)
+	y := EventController[Bar]{}
+	println(y)
+	assert true
+}

